### ISSUE TITLE
New Rule: max-complexity

### ISSUE THREAD

**Issue by pertrai1** [pertrai1/eslint-plugin-llm-core#114]
## Problem

LLMs generate functions with high cyclomatic complexity — long `if/else` chains, nested conditionals, and switch statements with many cases. ESLint's built-in `complexity` rule catches this, but its error message is opaque: `"Function 'foo' has a complexity of 12. Maximum allowed is 10."` It tells you *what* but not *why* or *how to fix it*. LLMs receiving that message often just shuffle the conditionals around without actually decomposing the logic.

This plugin's value proposition is teaching messages that LLMs can parse and apply in one pass. A `max-complexity` rule with structured teaching messages would fill a real gap — the README currently recommends `complexity: ["error", 10]` as a complementary rule, but users must configure it separately and lose the LLM-oriented guidance. By shipping our own, we replace that recommendation with a first-class rule that provides actionable fix guidance.

## Rule

Enforce a maximum cyclomatic complexity per function, matching ESLint's built-in `complexity` rule (classic variant) for counting semantics, with teaching-oriented error messages that explain decomposition strategies.

## Counting Contract (Frozen)

This rule uses **ESLint classic complexity parity** — the same nodes that ESLint's built-in `complexity` rule counts, using the same `onCodePathStart`/`onCodePathEnd` boundary semantics. Base complexity starts at 1. The following AST nodes each increment complexity by 1:

| Node | What counts | What does NOT count |
|---|---|---|
| `IfStatement` | Every `if` and every `else if` (each is a separate `IfStatement` node) | Plain `else` blocks (not an `IfStatement`) |
| `ForStatement`, `ForInStatement`, `ForOfStatement` | All loop types | — |
| `WhileStatement`, `DoWhileStatement` | All loop types | — |
| `CatchClause` | `catch` blocks | — |
| `ConditionalExpression` | Ternary `? :` (each ternary counts, including nested) | — |
| `LogicalExpression` | `&&`, `||`, `??` | — |
| `SwitchCase[test]` | Each non-default case (classic mode) | `default` case |
| `AssignmentPattern` | Default parameter values | — |
| `AssignmentExpression` with logical operator | `&&=`, `||=`, `??=` | Regular `=` assignments |
| `MemberExpression(optional: true)` | Optional chaining `?.` on property access | Non-optional member access |
| `CallExpression(optional: true)` | Optional chaining `?.()` on method calls | Non-optional calls |

**Per-function boundary:** Complexity resets at each code path boundary — functions, class field initializers, and class static blocks are each evaluated independently. A function containing a nested helper does not inherit the helper's complexity.

**Deliberate scoping:** This rule does **not** support ESLint's `modified` complexity variant. Classic mode only.

## Examples: Incorrect

```ts
// Cyclomatic complexity = 11 (exceeds default threshold of 10)
// Counted: base 1 + 8 IfStatements + 1 LogicalExpression (&&) + 1 more IfStatement = 11
function getDiscount(customerType: string, tier: number, isHoliday: boolean, isMember: boolean): number {
  if (customerType === "premium") {
    if (tier === 1) {
      if (isHoliday && isMember) {
        return 0.35;
      } else if (isHoliday) {
        return 0.30;
      } else if (isMember) {
        return 0.25;
      } else {
        return 0.20;
      }
    } else if (tier === 2) {
      if (isHoliday) {
        return 0.20;
      } else {
        return 0.15;
      }
    } else {
      return 0.10;
    }
  } else if (customerType === "standard") {
    if (isHoliday) {
      return 0.10;
    } else {
      return 0.05;
    }
  } else {
    return 0;
  }
}
```

```ts
// Long if/else chain — complexity = 7 (below default 10, but fails with max: 6)
// Counted: base 1 + 6 IfStatements (1 if + 5 else if)
function getStatusCode(error: AppError): number {
  if (error.type === "validation") return 400;
  else if (error.type === "auth") return 401;
  else if (error.type === "forbidden") return 403;
  else if (error.type === "notFound") return 404;
  else if (error.type === "conflict") return 409;
  else if (error.type === "rateLimit") return 429;
  else return 500;
}
```

## Examples: Correct

```ts
// Lookup table — complexity = 2 (base 1 + ?? LogicalExpression)
const ERROR_STATUS: Record<string, number> = {
  validation: 400,
  auth: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  rateLimit: 429,
};

function getStatusCode(error: AppError): number {
  return ERROR_STATUS[error.type] ?? 500;
}
```

```ts
// Strategy pattern with data object — complexity = 4 (base 1 + optional chaining + 2 ConditionalExpressions)
const DISCOUNT_RATES: Record<string, Record<number, { holiday: number; normal: number }>> = {
  premium: {
    1: { holiday: 0.30, normal: 0.25 },
    2: { holiday: 0.20, normal: 0.15 },
  },
  standard: {
    1: { holiday: 0.10, normal: 0.05 },
  },
};

function getDiscount(customerType: string, tier: number, isHoliday: boolean): number {
  const rates = DISCOUNT_RATES[customerType]?.[tier];
  return rates ? (isHoliday ? rates.holiday : rates.normal) : 0;
}
```

Both "correct" examples are dramatically simpler than the `if/else` versions — even though they are not complexity 1, they are well within any reasonable threshold.

**Note:** All complexity counts above verified against ESLint's built-in `complexity` rule using `npx eslint --rule 'complexity:["error",0]'`.

## Why it matters

High cyclomatic complexity is the #1 predictor of bugs in LLM-generated code. Each branching path is an independent surface where the LLM can introduce a logic error. At complexity 6+, the LLM's attention mechanism starts "losing" earlier branches — it contradicts itself, duplicates logic, or misses edge cases.

ESLint's built-in `complexity` detects this but provides no remediation guidance. An LLM receiving `"Function has complexity of 12"` doesn't know whether to use a lookup table, strategy pattern, or guard clauses. A teaching message that names a concrete decomposition strategy — with a before/after example — enables one-pass fixes instead of trial-and-error refactoring.

## Scope boundaries

**In scope (flag):**
- Functions (declarations, expressions, arrows) where cyclomatic complexity exceeds the configured threshold
- Class field initializers and class static blocks (matching ESLint behavior)
- All node types listed in the Counting Contract table above
- Per-function boundary: complexity resets at each code path boundary

**Out of scope (do NOT flag):**
- Files or modules as a whole (this is a per-function metric)
- Class-level complexity (sum of method complexities) — too noisy
- ESLint's `modified` complexity variant (classic mode only)

**False-positive boundaries:**
- Small switch statements with 3-5 simple cases are not inherently complex. The default threshold of 10 accommodates these. Users who want stricter limits can configure lower.
- Plain `else` blocks do NOT count as complexity increments — only `IfStatement` nodes do. This matches ESLint's built-in behavior.
- `default` cases in switch statements do NOT count in classic mode.

## Config Interaction with ESLint's Built-in `complexity`

The README currently recommends core ESLint `complexity` as a complementary rule. When this rule ships:

1. **Remove** the `complexity` row from the README's "Complementary Rules" table.
2. **Replace** it with a note in `max-complexity`'s docs: "If you previously used ESLint's `complexity` rule, you can remove it — this rule uses the same counting semantics with added teaching messages."
3. The `recommended` config will include `llm-core/max-complexity` at the default threshold. Users who also have ESLint's `complexity` enabled will get duplicate diagnostics — the docs should note this and recommend removing the ESLint rule.

## Proposed config placement

**`recommended`** — High complexity is universally harmful, and the teaching messages are the primary value. Every project benefits from this check.

Also added to the **`complexity`** config alongside `max-file-length`, `max-function-length`, `max-nesting-depth`, and `max-params`.

## Framework-agnostic proof

Cyclomatic complexity is a language-level metric that applies to any TypeScript/JavaScript function regardless of framework, architecture, or directory layout. The detection logic operates purely on AST control flow nodes. No framework-specific knowledge is required.

## Prior art

- **ESLint built-in `complexity`** — Detects the same metric using `onCodePathStart`/`onCodePathEnd`. This rule uses identical counting semantics (classic variant) but with structured error messages following the plugin's message template convention. The implementation should use the same code-path-based approach rather than manual node counting.
- **`max-nesting-depth`** (this plugin) — Measures *nesting depth*, not *branching count*. A flat if/else chain with 10 branches has nesting depth 1 but complexity 10. These are complementary metrics.
- **`prefer-early-return`** (this plugin) — Catches the specific case of a function body wrapped in a single if. Does not address overall branching complexity.
- **`max-function-length`** (this plugin) — Measures lines, not paths. A 50-line function with no branching has complexity 1. A 10-line function with 8 ternaries can have complexity 9. Complementary.

## Options

```ts
type Options = [
  {
    max?: number;            // Default: 10
    skipTestFiles?: boolean; // Default: true
  },
];
```

---

## Proposed Error Message (Frozen)

This message follows the repo's lint-message convention per the active decision log (`docs/decisions/2026-04-05-lint-message-template.md`): one direct violation sentence, one short `Why:` line, and a concrete `How to fix:` section anchored in `Before:` and `After:` rewrites.

```
Function '{{ name }}' has a complexity of {{ count }}, exceeding the maximum of {{ max }}.

Why: Each branching path is an independent surface for logic errors — high-complexity functions are harder to test and modify safely.

How to fix:
  Replace branching logic with a data structure.
  Before: if (type === 'a') return 1; else if (type === 'b') return 2; else if (type === 'c') return 3;
  After:  const VALUES: Record<string, number> = { a: 1, b: 2, c: 3 };
          function getValue(type: string): number { return VALUES[type] ?? 0; }
```

**Design note:** The message shows one strategy (lookup table) with one before/after, matching the convention enforced by `tests/rules/message-guidance.test.ts`. The lookup table is the most broadly applicable fix for the most common LLM complexity pattern (value-mapping if/else chains). Other strategies (guard clauses for nesting, function extraction for mixed concerns) are documented in `docs/rules/max-complexity.md` rather than in the lint message itself. This keeps the message concise while the doc provides full guidance.

---

## Definition of Done

- [ ] Rule file: `src/rules/max-complexity.ts` using `createRule`
- [ ] Implementation uses `onCodePathStart`/`onCodePathEnd` (matching ESLint's approach) rather than manual node counting
- [ ] Exported from `src/rules/index.ts` (alphabetical)
- [ ] Added to `complexityRules` and `recommendedRules` in `src/index.ts`
- [ ] Teaching message follows the repo's frozen message convention (one `Why:` line, concrete `Before:`/`After:` rewrites)
- [ ] Message passes `expectSingleWhyLine` and `expectTemplateShape` from `tests/rules/message-guidance.test.ts`
- [ ] Cyclomatic complexity calculation matches ESLint's built-in `complexity` rule (classic variant, all nodes listed in Counting Contract table)
- [ ] Per-function boundary: complexity resets at each code path boundary (functions, class field initializers, class static blocks)
- [ ] `skipTestFiles` option: when true (default), the rule skips the **entire file** if the filename matches `*.test.*` or `*.spec.*` — this includes functions, class field initializers, and class static blocks within that file. Matches the pattern used by `max-function-length`.
- [ ] Tests: `tests/rules/max-complexity.test.ts` covering:
  - Functions at threshold (pass)
  - Functions above threshold (fail, correct message)
  - Each complexity-increasing node type:
    - `IfStatement` (if, else if — NOT plain else)
    - `ForStatement`, `ForInStatement`, `ForOfStatement`
    - `WhileStatement`, `DoWhileStatement`
    - `CatchClause`
    - `ConditionalExpression` (ternary, including nested)
    - `LogicalExpression` (`&&`, `||`, `??`)
    - `SwitchCase[test]` (non-default cases only in classic mode)
    - `AssignmentPattern` (default parameters)
    - Logical assignment operators (`&&=`, `||=`, `??=`)
    - Optional member access (`?.`)
    - Optional call (`?.()`)
  - Per-function boundary reset (nested function does not inherit parent's complexity)
  - Class field initializer boundary
  - Class static block boundary
  - `skipTestFiles` skips entire file (functions, class field initializers, and static blocks in test files all skipped)
  - Custom threshold option
  - Arrow functions and method definitions
  - Anonymous functions resolve to variable name when possible
- [ ] Docs: `docs/rules/max-complexity.md` (includes all decomposition strategies, not just the one in the message)
- [ ] README updated: remove `complexity` from Complementary Rules table, add migration note
- [ ] `npm run update:eslint-docs` run
- [ ] `npm test && npm run lint && npm run build` all pass
- [ ] RED/GREEN/REFACTOR cycle followed — `test:` commits before `feat:` commits

**Comment by pertrai1** [comment]
I don’t think is implementation-ready yet. The main discrepancies are:

- The first “Incorrect” example contradicts the stated default. The issue says `getDiscount` has complexity `8` with a default threshold of `10`, but that should be valid, not invalid. If the rule is implemented as described, the examples and tests will disagree on day one.

- The “match ESLint built-in `complexity`” requirement is underspecified and partly wrong. ESLint’s rule increments for `IfStatement`, loops, `CatchClause`, `ConditionalExpression`, `LogicalExpression`, `AssignmentPattern`, logical assignment operators, and optional chaining/member calls, and it excludes `default` switch cases in classic mode; it does not count plain `else` as its own increment. That behavior is visible in [node_modules/eslint/lib/rules/complexity.js](/Users/robsimpson/Repos/eslint-plugin-llm-core/node_modules/eslint/lib/rules/complexity.js:119). The issue’s scope section currently says `else` counts, and it omits `AssignmentPattern`, optional chaining, and logical assignment operators.

- The “Correct” examples also conflict with built-in semantics. `return ERROR_STATUS[error.type] ?? 500;` is not complexity `1`; `??` adds a branch, so it is at least `2`. The “strategy pattern” example also uses optional chaining and ternaries, so it is not `1` either. That means the issue currently teaches the wrong contract, which the repo explicitly warns against in the active lint-message decision log at [docs/decisions/2026-04-05-lint-message-template.md](/Users/robsimpson/Repos/eslint-plugin-llm-core/docs/decisions/2026-04-05-lint-message-template.md:27).

- The config story is unresolved. The README still recommends core ESLint `complexity` as a complementary rule in [README.md](/Users/robsimpson/Repos/eslint-plugin-llm-core/README.md:181). If `llm-core/max-complexity` is added to `recommended`, the issue should explicitly say whether the README guidance will be removed or replaced, otherwise users will likely enable both and get duplicate diagnostics for the same function.

- The message requirements are still partly “for discussion,” which means the acceptance target is not frozen. The repo’s current message contract is “one direct violation sentence, one short `Why:` line, and concrete `Before:` / `After:` rewrites” per [tests/rules/message-guidance.test.ts](/Users/robsimpson/Repos/eslint-plugin-llm-core/tests/rules/message-guidance.test.ts:10) and the active decision log at [docs/decisions/2026-04-05-lint-message-template.md](/Users/robsimpson/Repos/eslint-plugin-llm-core/docs/decisions/2026-04-05-lint-message-template.md:27). The proposed message is much longer and leaves open whether the rule should stay static or become pattern-specific.

- The Definition of Done is incomplete if parity with ESLint is truly required. The test list omits default parameters, optional chaining, logical assignment operators, and the built-in rule’s class-field-initializer / class-static-block edge cases from [node_modules/eslint/lib/rules/complexity.js](/Users/robsimpson/Repos/eslint-plugin-llm-core/node_modules/eslint/lib/rules/complexity.js:156). Those cases need either explicit tests or an explicit out-of-scope statement.

What would make it ready is a short requirements edit pass:
1. Fix the invalid/valid examples and their claimed counts.
2. Freeze the exact counting contract as “ESLint classic complexity parity” or narrow it deliberately.
3. Decide how this interacts with core ESLint `complexity` in docs/configs.
4. Freeze the message shape to match the repo’s current lint-message convention.
5. Expand the DoD test matrix to cover all counted node types, or explicitly exclude the omitted ones.

No code changes made.

**Comment by pertrai1** [comment]
**Findings**

- [P1] The example counts still do not match the frozen “ESLint classic parity” contract in [issue #114](https://github.com/pertrai1/eslint-plugin-llm-core/issues/114). I verified them against ESLint’s actual `complexity` rule in [node_modules/eslint/lib/rules/complexity.js](/Users/robsimpson/Repos/eslint-plugin-llm-core/node_modules/eslint/lib/rules/complexity.js:113). The first “Incorrect” example reports complexity `11`, not `12`, and the second “Incorrect” example reports `7`, not `8`. The two “Correct” examples are now accurate at `2` and `4`. This is still a real implementation blocker because tests and docs derived from those examples would be off by one.

How to reproduce:

```
tmp=./max-complexity-$(date +%s%N).js
cat > "$tmp" <<'EOF'
function getDiscount(customerType, tier, isHoliday, isMember) {
  if (customerType === "premium") {
    if (tier === 1) {
      if (isHoliday && isMember) {
        return 0.35;
      } else if (isHoliday) {
        return 0.30;
      } else if (isMember) {
        return 0.25;
      } else {
        return 0.20;
      }
    } else if (tier === 2) {
      if (isHoliday) {
        return 0.20;
      } else {
        return 0.15;
      }
    } else {
      return 0.10;
    }
  } else if (customerType === "standard") {
    if (isHoliday) {
      return 0.10;
    } else {
      return 0.05;
    }
  } else {
    return 0;
  }
}
EOF
npx eslint "$tmp" --no-config-lookup --rule 'complexity:["error",0]'
rm "$tmp"
```

- [P2] `skipTestFiles` is slightly underspecified now that the rule explicitly includes class field initializers and class static blocks. The issue says test files are out of scope via `skipTestFiles`, but it phrases that section in terms of “functions in test files.” If the intended behavior is “skip the entire file,” matching the current pattern used by rules like [max-function-length](/Users/robsimpson/Repos/eslint-plugin-llm-core/src/rules/max-function-length.ts:77), say that explicitly so there is no ambiguity about whether class field initializers/static blocks in `*.test.*` files are skipped too.

How to reproduce:

```
tmp=./max-complexity-$(date +%s%N).js
cat > "$tmp" <<'EOF'
function getStatusCode(error) {
  if (error.type === "validation") return 400;
  else if (error.type === "auth") return 401;
  else if (error.type === "forbidden") return 403;
  else if (error.type === "notFound") return 404;
  else if (error.type === "conflict") return 409;
  else if (error.type === "rateLimit") return 429;
  else return 500;
}
EOF
npx eslint "$tmp" --no-config-lookup --rule 'complexity:["error",0]'
rm "$tmp"
```
